OVA upgrade from 1.2.x to 1.2.y

This issue is for delivering an OVA upgrade procedure from 1.2.x to 1.2.y, which is different from 1.1.1 to 1.2.x since data migration for Admiral is not involved.

**Please see https://github.com/vmware/vic-product/issues/839#issuecomment-330950081 for more details. Leaving the original issue details below for posterity.**

---

/cc @sergiosagu @gmuleshkov @mdubya66 

**Data migration script:**

https://github.com/vmware/admiral/blob/master/upgrade/src/main/resources/migrate.sh

**Upgrading from:**

https://storage.googleapis.com/vic-product-ova-builds/vic-v1.2.1-rc1-15e8cde5.ova
(Post 1.2.0 build containing Admiral v1.2.1)

To test data migration, this OVA instance had (all actions performed through Admiral):

* an additional project, `project-2`
* the `busybox` image pushed to the OVA's Harbor repository for both projects
* a VCH as a Container Host with one `busybox` container running `/bin/sh` in `default-project` 

**Upgrading to:**

https://storage.googleapis.com/vic-product-ova-builds/vic-5b3261d8-devrc3.ova
(Post 1.2.1-rc2 build containing Admiral v1.2.1, Harbor 1.2.0 and Engine 1.2.1-rc2. Contains fix for OVA upgrade with external PSC: https://github.com/vmware/vic-product/pull/838)

**Observed behavior:**

Following the upgrade procedure from https://vmware.github.io/vic-product/assets/files/html/1.2/vic_vsphere_admin/upgrade_appliance.html, step 12:

```
# ./upgrade_1.1_to_1.2.sh 
Getting password from harbor.cfg
Enter vCenter Server FQDN or IP: <vCenter>
Enter vCenter Administrator Username: <user>
Enter vCenter Administrator Password: <password>
If using an external PSC, enter the FQDN of the PSC instance (leave blank otherwise): 
If using an external PSC, enter the PSC Admin Domain (leave blank otherwise): 
Starting upgrade 2017-09-19 22:50:47 +0000 UTC
Preparing upgrade environment
Disabling and stopping Admiral and Harbor path startup
Finished preparing upgrade environment
Performing pre-upgrade checks
Starting Admiral upgrade
[=] Shutting down Harbor and Admiral
[=] Upgrade Admiral
Data migration to new Admiral failed
```

**Debugging:**

The `migrate.sh` script references a `$auth_token` variable in two places:
* https://github.com/vmware/admiral/blob/master/upgrade/src/main/resources/migrate.sh#L20 and
* https://github.com/vmware/admiral/blob/master/upgrade/src/main/resources/migrate.sh#L26. 

I didn’t find where this variable was defined. This could be a typo - I see `$3` (the correct token input parameter) used in other places in this file, should we be using the same token value here?

I changed `$auth_token` to `$3` but that didn’t fix things - this could be a red herring since I'm not fully aware of how the `migrate.sh` script's environment is set up, but I think it's worth mentioning.

I then ran the `migrate.sh` script manually with the parameters it’s invoked with in the [upgrade script](https://github.com/vmware/vic-product/blob/master/installer/packer/scripts/upgrade/upgrade_1.1_to_1.2.sh):

```
# /etc/vmware/admiral/migrate.sh 10.161.67.28:8283 10.161.67.28:8282 $psc_token
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
Old admiral node is ready for migration: 10.161.67.28:8283
New admiral node is ready for migration: 10.161.67.28:8282
Starting migration of the data
<... nothing for 5-10 mins ...>
Status code from migration: 500
Error during migration please check xenon logs for more information!
```

The script fails here: https://github.com/vmware/admiral/blob/master/upgrade/src/main/resources/migrate.sh#L59

**Summary:**

I’m not sure how to check xenon logs here. The only other thing I can think of is that **this is a 1.2.x -> 1.2.y upgrade, not a 1.1.x to 1.2.y upgrade**. Are there expectations around data schema/format during data migration? Does the migration code expect the data to follow the data schema/format from 1.1.1? In the upgrade script, I see we’re launching an Admiral 1.1.1 container to upgrade from, regardless of the version of Admiral in the older OVA (1.2.0 in this case): https://github.com/vmware/vic-product/blob/master/installer/packer/scripts/upgrade/upgrade_1.1_to_1.2.sh#L281

Since the upgrade process for `1.2.x to 1.2.y` will be different from that of `1.1.1 to 1.2.x`, I attempted the following upgrade procedure on a fresh set of `1.2.0 GA` and `1.2.1-dev-rc3` OVAs:

* Followed the [usual OVA upgrade steps](https://vmware.github.io/vic-product/assets/files/html/1.2/vic_vsphere_admin/upgrade_appliance.html), except for running the upgrade script (step 12).
* After powering on new OVA, instead of running the upgrade script, simply go to the Getting Started page (:9443) and log in.
* Old OVA’s data (projects, container hosts) can be seen in the new Admiral's UI.

**However, Harbor data and UI elements look broken. Also, after getting the red login prompt at the top of the Admiral UI and logging in again, project data looks inconsistent and takes a while to load.** 

<img width="810" alt="registry_conf" src="https://user-images.githubusercontent.com/4361620/30662390-83937422-9e0c-11e7-8414-64df4d980afc.png">
<img width="1414" alt="registry_repos" src="https://user-images.githubusercontent.com/4361620/30662391-839a7c86-9e0c-11e7-8967-06b5473ba6a5.png">
<img width="837" alt="project_data" src="https://user-images.githubusercontent.com/4361620/30662418-9c59c3c6-9e0c-11e7-921f-3d4cfc289844.png">


I see this exception in the Admiral container logs when loading registry data:

```
[473][W][2017-09-20T18:22:31.322Z][56][8282/][lambda$retrieveRepositoriesAndTagsCount$36][Could not retrieve repositories for project /projects/default-project with harborId 1: java.util.concurrent.CompletionException: java.lang.IllegalStateException: No route to host: /10.192.130.1:443
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:604)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at com.vmware.xenon.common.DeferredResult.complete(DeferredResult.java:402)
	at com.vmware.xenon.common.ServiceRequestSender.lambda$sendWithDeferredResult$1(ServiceRequestSender.java:48)
	at com.vmware.xenon.common.Operation.lambda$nestCompletion$1(Operation.java:1348)
	at com.vmware.xenon.common.Operation.completeOrFail(Operation.java:1317)
	at com.vmware.xenon.common.Operation.complete(Operation.java:1219)
	at com.vmware.admiral.service.common.harbor.HarborApiProxyService.lambda$forwardRequest$3(HarborApiProxyService.java:168)
	at com.vmware.xenon.common.Operation.completeOrFail(Operation.java:1317)
	at com.vmware.xenon.common.Operation.fail(Operation.java:1295)
	at com.vmware.xenon.common.Operation.fail(Operation.java:1223)
	at com.vmware.xenon.common.http.netty.NettyHttpServiceClient.fail(NettyHttpServiceClient.java:725)
	at com.vmware.xenon.common.http.netty.NettyHttpServiceClient.lambda$connectChannel$0(NettyHttpServiceClient.java:421)
	at com.vmware.xenon.common.Operation.lambda$nestCompletion$1(Operation.java:1348)
	at com.vmware.xenon.common.Operation.completeOrFail(Operation.java:1317)
	at com.vmware.xenon.common.Operation.fail(Operation.java:1295)
	at com.vmware.xenon.common.http.netty.NettyChannelPool.fail(NettyChannelPool.java:579)
	at com.vmware.xenon.common.http.netty.NettyChannelPool.lambda$connectOrReuse$1(NettyChannelPool.java:356)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:507)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:500)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:479)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:420)
	at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:122)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.fulfillConnectPromise(AbstractNioChannel.java:327)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:343)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:632)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: No route to host: /10.192.130.1:443
	at com.vmware.admiral.auth.util.ProjectUtil.lambda$retrieveRepositoriesAndTagsCount$35(ProjectUtil.java:426)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	... 34 more
]
[474][W][2017-09-20T18:22:31.329Z][58][8282/][lambda$retrieveRepositoriesAndTagsCount$36][Could not retrieve repositories for project /projects/fffb49187d1d8e75559a1a4f1b158 with harborId 1419277516: java.util.concurrent.CompletionException: java.lang.IllegalStateException: No route to host: /10.192.130.1:443
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:604)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at com.vmware.xenon.common.DeferredResult.complete(DeferredResult.java:402)
	at com.vmware.xenon.common.ServiceRequestSender.lambda$sendWithDeferredResult$1(ServiceRequestSender.java:48)
	at com.vmware.xenon.common.Operation.lambda$nestCompletion$1(Operation.java:1348)
	at com.vmware.xenon.common.Operation.completeOrFail(Operation.java:1317)
	at com.vmware.xenon.common.Operation.complete(Operation.java:1219)
	at com.vmware.admiral.service.common.harbor.HarborApiProxyService.lambda$forwardRequest$3(HarborApiProxyService.java:168)
	at com.vmware.xenon.common.Operation.completeOrFail(Operation.java:1317)
	at com.vmware.xenon.common.Operation.fail(Operation.java:1295)
	at com.vmware.xenon.common.Operation.fail(Operation.java:1223)
	at com.vmware.xenon.common.http.netty.NettyHttpServiceClient.fail(NettyHttpServiceClient.java:725)
	at com.vmware.xenon.common.http.netty.NettyHttpServiceClient.lambda$connectChannel$0(NettyHttpServiceClient.java:421)
	at com.vmware.xenon.common.Operation.lambda$nestCompletion$1(Operation.java:1348)
	at com.vmware.xenon.common.Operation.completeOrFail(Operation.java:1317)
	at com.vmware.xenon.common.Operation.fail(Operation.java:1295)
	at com.vmware.xenon.common.http.netty.NettyChannelPool.fail(NettyChannelPool.java:579)
	at com.vmware.xenon.common.http.netty.NettyChannelPool.lambda$connectOrReuse$1(NettyChannelPool.java:356)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:507)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:500)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:479)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:420)
	at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:122)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.fulfillConnectPromise(AbstractNioChannel.java:327)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:343)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:632)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: No route to host: /10.192.130.1:443
	at com.vmware.admiral.auth.util.ProjectUtil.lambda$retrieveRepositoriesAndTagsCount$35(ProjectUtil.java:426)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	... 34 more
]
```

`No route to host: /10.192.130.1:443`: `10.192.130.1:443` is the old address of Harbor on the old OVA.

Looks like Admiral's data is fine, but the Harbor address stored in Admiral needs to be updated, plus any other Harbor-related changes. These are handled in the [upgrade script](https://github.com/vmware/vic-product/blob/master/installer/packer/scripts/upgrade/upgrade_1.1_to_1.2.sh#L389). Based on this, my understanding is that we need only the Harbor-related upgrade pieces (with modifications) from the [upgrade script](https://github.com/vmware/vic-product/blob/master/installer/packer/scripts/upgrade/upgrade_1.1_to_1.2.sh) for `1.2.x -> 1.2.y`.

--- a/installer/packer/scripts/upgrade/upgrade.sh
+++ b/installer/packer/scripts/upgrade/upgrade.sh
@@ -350,7 +350,7 @@ function upgradeAdmiral {
   if [ -n "$(docker ps -q -f name=vic-upgrade-admiral)" ]; then
     echo "Admiral upgrade container already exists" | tee /dev/fd/3
     echo "If upgrade is not already running, execute the following command and rerun the upgrade script:" | tee /dev/fd/3
-    echo "    docker rm -f vic-upgrade-admrial" | tee /dev/fd/3
+    echo "    docker rm -f vic-upgrade-admiral" | tee /dev/fd/3
     exit 1
   fi
 
@@ -580,6 +580,14 @@ function main {
   echo "-------------------------"
   echo "Starting upgrade ${TIMESTAMP}" | tee /dev/fd/3
 
+  data_upgrade_needed=false
+  if [ ! -f "/data/admiral/configs/psc-config.properties" ]; then
+      echo "Detected old OVA's version as 1.1.x. Continuing upgrade operation with data migration." | tee /dev/fd/3
+      data_upgrade_needed=true
+  else
+      echo "Detected old OVA's version as 1.2.x. Continuing upgrade operation without data migration." | tee /dev/fd/3
+  fi
+
   echo "Preparing upgrade environment" | tee /dev/fd/3
   disableServicesStart
   registerAppliance
@@ -587,9 +595,25 @@ function main {
   writeTimestamp
   echo "Finished preparing upgrade environment" | tee /dev/fd/3
 
-  upgradeAdmiral
+  if [ "$data_upgrade_needed" = true ]; then
+      upgradeAdmiral
+  else
+      # Start Admiral
+      echo "Starting Admiral" | tee /dev/fd/3
+      systemctl start admiral_startup.service
+      sleep 3
+  fi
+
   updateAdmiralConfig
-  upgradeHarbor
+
+  if [ "$data_upgrade_needed" = true ]; then
+      upgradeHarbor
+  else
+      # Start Harbor
+      echo ""
+      echo "Starting Harbor" | tee /dev/fd/3
+      systemctl start harbor_startup.service
+  fi
   setDataVersion
 
   enableServicesStart

--- a/installer/packer/scripts/upgrade/upgrade.sh
+++ b/installer/packer/scripts/upgrade/upgrade.sh
@@ -22,6 +22,8 @@ harbor_psc_token_file="/etc/vmware/psc/harbor/tokens.properties"
 admiral_psc_token_file="/etc/vmware/psc/admiral/tokens.properties"
 timestamp_file="/registration-timestamps.txt"
 
+data_upgrade_needed=false
+
 admiral_upgrade_status="/etc/vmware/admiral/upgrade_status"
 harbor_upgrade_status="/etc/vmware/harbor/upgrade_status"
 upgrade_log_file="/var/log/vmware/upgrade.log"
@@ -494,6 +496,62 @@ function enableServicesStart {
   systemctl start harbor_startup.path
 }
 
+# Check for presence of Admiral's PSC config file. If the file exists, the old
+# OVA is version 1.2.x, otherwise it is 1.1.x and data migration is needed.
+function setDataUpgradeNeeded {
+  if [ ! -f "/data/admiral/configs/psc-config.properties" ]; then
+    echo "Detected old OVA's version as 1.1.x. Upgrade will perform data migration." | tee /dev/fd/3
+
+      while true; do
+        echo "" | tee /dev/fd/3
+        echo "Do you wish to proceed with a 1.1.x to 1.2.y upgrade? [y/n]" | tee /dev/fd/3
+        echo -n "If the version of the old OVA is not 1.1.x, please enter n and contact VMware support: " | tee /dev/fd/3
+        read response
+        case $response in
+            [Yy] )
+                echo "Continuing with upgrade" | tee /dev/fd/3
+                echo "" | tee /dev/fd/3
+                break
+                ;;
+            [Nn] )
+                echo "Exiting without performing upgrade" | tee /dev/fd/3
+                exit 1
+                ;;
+            *)
+                # unknown option
+                echo "Please enter [y/n]" | tee /dev/fd/3
+                ;;
+        esac
+      done
+
+      data_upgrade_needed=true
+  else
+    echo "Detected old OVA's version as 1.2.x. Upgrade will not perform data migration." | tee /dev/fd/3
+
+      while true; do
+        echo "" | tee /dev/fd/3
+        echo "Do you wish to proceed with a 1.2.x to 1.2.y upgrade? [y/n]" | tee /dev/fd/3
+        echo -n "If the version of the old OVA is not 1.2.x, please enter n and contact VMware support: " | tee /dev/fd/3
+        read response
+        case $response in
+            [Yy] )
+                echo "Continuing with upgrade" | tee /dev/fd/3
+                echo "" | tee /dev/fd/3
+                break
+                ;;
+            [Nn] )
+                echo "Exiting without performing upgrade" | tee /dev/fd/3
+                exit 1
+                ;;
+            *)
+                # unknown option
+                echo "Please enter [y/n]" | tee /dev/fd/3
+                ;;
+        esac
+      done
+  fi
+}
+
 function main {
   while [[ $# -gt 1 ]]
   do
@@ -580,13 +638,7 @@ function main {
   echo "-------------------------"
   echo "Starting upgrade ${TIMESTAMP}" | tee /dev/fd/3
 
-  data_upgrade_needed=false
-  if [ ! -f "/data/admiral/configs/psc-config.properties" ]; then
-      echo "Detected old OVA's version as 1.1.x. Continuing upgrade operation with data migration." | tee /dev/fd/3
-      data_upgrade_needed=true
-  else
-      echo "Detected old OVA's version as 1.2.x. Continuing upgrade operation without data migration." | tee /dev/fd/3
-  fi
+  setDataUpgradeNeeded
 
   echo "Preparing upgrade environment" | tee /dev/fd/3
   disableServicesStart

